index_release.go

```diff
@@ -19,6 +20,95 @@ type IndexRelease struct {
 	Version     string      `yaml:"version"`
 }
 
+// CompileReleases takes indexReleases and collected version bundles and
+// compiles canonicalized Releases from them.
+func CompileReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) ([]Release, error) {
+	releases := buildReleases(logger, indexReleases, bundles)
+
+	deduplicateReleaseChangelog(releases)
+
+	return releases, nil
+}
+
+func buildReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) []Release {
+	bundleCache := make(map[string]Bundle)
+
+	// Create cache of bundles for quick lookup
+	for _, b := range bundles {
+		bundleCache[b.ID()] = b
+	}
+
+	var releases []Release
+
+INDEX_RELEASES:
+	for _, ir := range indexReleases {
+		release := Release{
+			active:    ir.Active,
+			timestamp: ir.Date.Format(indexReleaseTimestampFormat),
+			version:   ir.Version,
+		}
+
+		for _, a := range ir.Authorities {
+			b, found := bundleCache[a.BundleID()]
+			if !found {
+				logger.Log("level", "warning", "message", "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles. Skipping.")
+				continue INDEX_RELEASES
+			}
+			release.bundles = append(release.bundles, b)
+		}
```

Comment:
Can we have some function `groupBundlesForIndexRelease(ir, bundles)`?  It can return the list of grouped bundles and a not found error in case some bundle is missing. That way we get rid of the loop label. I hate these. :D IMO the code would be easier to understand then and we can assign the returned list or do the logging separated from the grouping logic.